subsys/net/ip/tcp2.c

```diff
@@ -267,6 +274,145 @@ int net_tcp_unref(struct net_context *context)
 	return ref_count;
 }
 
+static int tcp_finalize_pkt(struct net_pkt *pkt)
+{
+	net_pkt_cursor_init(pkt);
+
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_ipv4_finalize(pkt, IPPROTO_TCP);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_ipv6_finalize(pkt, IPPROTO_TCP);
+	}
+
+	return -EINVAL;
+}
+
+static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
+{
+	NET_PKT_DATA_ACCESS_DEFINE(tcp_access, struct tcphdr);
+	struct tcphdr *th;
+
+	th = (struct tcphdr *)net_pkt_get_data(pkt, &tcp_access);
+	if (!th) {
+		return -ENOBUFS;
+	}
+
+	memset(th, 0, sizeof(struct tcphdr));
+
+	th->th_sport = conn->src.sin.sin_port;
+	th->th_dport = conn->dst.sin.sin_port;
+
+	th->th_off = 5;
+	th->th_flags = flags;
+	th->th_win = htons(conn->recv_win);
+	th->th_seq = htonl(conn->seq);
+
+	if (ACK & flags) {
+		th->th_ack = htonl(conn->ack);
+	}
+
+	return net_pkt_set_data(pkt, &tcp_access);
+}
+
+static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+{
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_context_create_ipv4_new(conn->context, pkt,
+						&conn->src.sin.sin_addr,
+						&conn->dst.sin.sin_addr);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_context_create_ipv6_new(conn->context, pkt,
+						&conn->src.sin6.sin6_addr,
+						&conn->dst.sin6.sin6_addr);
+	}
+
+	return -EINVAL;
+}
+
+static void tcp_send_queued_data(struct k_work *work)
+{
+	struct tcp *conn = CONTAINER_OF(work, struct tcp, send_timer);
+	struct net_pkt *pkt;
+	int len;
+	int ret;
+
+	len = net_pkt_get_len(conn->pending_data) - conn->pending_len;
+	if (len < 0) {
+		NET_DBG("Packet %p len %zd is shorter than pending data %zd",
+			conn->pending_data,
+			net_pkt_get_len(conn->pending_data),
+			conn->pending_len);
+		return;
+	}
+
+	/* TODO: We should adjust the amount of data to sent according to
+	 * our number of available buffers etc.
+	 */
+	len = MIN(conn->send_win, len);
+
+	NET_DBG("Sending %d win %zd pending %zd seq %u ack %u total %zd", len,
+		conn->send_win, conn->pending_len,
+		conn->seq, conn->ack, net_pkt_get_len(conn->pending_data));
+
+	pkt = net_pkt_alloc_with_buffer(conn->iface, len,
```

Comment:
Let's use tcp_pkt_alloc() here, so we can track it when the test protocol is enabled.

Reply:
Sure, that can be done.